File: packages/ui/bin/ui.js
#!/usr/bin/env node

/**
 * @leanspec/ui - Standalone web UI launcher
 * 
 * This is the entry point for the standalone @leanspec/ui package.
 * It wraps the @leanspec/web Next.js application and provides:
 * - CLI argument parsing
 * - Specs directory auto-detection
 * - Environment configuration for filesystem mode
 * - Browser auto-open
 */

import { spawn } from 'node:child_process';
import { existsSync } from 'node:fs';
import { resolve, join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { readFileSync } from 'node:fs';
import { load as loadYaml } from 'js-yaml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Parse command line arguments
 */
function parseArgs(args) {
  const options = {
    specs: null,
    port: '3000',
    open: true,
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    
    if (arg === '--specs' && i + 1 < args.length) {
      options.specs = args[++i];
    } else if (arg.startsWith('--specs=')) {
      options.specs = arg.slice('--specs='.length);
    } else if (arg === '--port' || arg === '-p') {
      if (i + 1 < args.length) {
        options.port = args[++i];
      }
    } else if (arg.startsWith('--port=')) {
      options.port = arg.slice('--port='.length);
    } else if (arg === '--no-open') {
      options.open = false;
    } else if (arg === '--help' || arg === '-h') {
      console.log(`
Usage: npx @leanspec/ui [options]

Options:
  --specs <dir>        Specs directory (auto-detected if not specified)
  --port, -p <port>    Port to run on (default: 3000)
  --no-open            Don't open browser automatically
  --help, -h           Show this help message

Examples:
  npx @leanspec/ui
  npx @leanspec/ui --specs ./my-specs
  npx @leanspec/ui --port 4000 --no-open
`);
      process.exit(0);
    }
  }

  return options;
}

/**
 * Auto-detect specs directory
 */
function detectSpecsDirectory(cwd) {
  const candidates = [
    './specs',
    './spec',
    './docs/specs',
    './docs/spec',
    './.lean-spec/specs'
  ];
  
  // Try standard locations first
  for (const candidate of candidates) {
    const fullPath = resolve(cwd, candidate);
    if (existsSync(fullPath)) {
      return fullPath;
    }
  }
  
  // Check for leanspec.yaml config
  const configPath = resolve(cwd, 'leanspec.yaml');
  if (existsSync(configPath)) {
    try {
      const config = loadYaml(readFileSync(configPath, 'utf-8'));
      if (config.specsDir) {
        return resolve(cwd, config.specsDir);
      }
    } catch (error) {
      // Ignore config parse errors
    }
  }
  
  return null;
}

/**
 * Validate port number
 */
function validatePort(port) {
  const portNum = parseInt(port, 10);
  if (isNaN(portNum) || portNum < 1 || portNum > 65535) {
    console.error(`✗ Invalid port number: ${port}`);
    console.log('Port must be between 1 and 65535');
    process.exit(1);
  }
  return portNum;
}

/**
 * Find the @leanspec/web package directory
 */
function findWebPackage() {
  // When installed via npm, @leanspec/web will be in node_modules
  const currentDir = dirname(__filename);
  
  // Try relative to this package (node_modules/@leanspec/ui/bin/ui.js)
  // -> node_modules/@leanspec/web
  const webInNodeModules = resolve(currentDir, '../../../@leanspec/web');
  if (existsSync(webInNodeModules)) {
    return webInNodeModules;
  }
  
  // Try in parent node_modules (for pnpm/yarn workspaces)
  const webInParentNodeModules = resolve(currentDir, '../../../../@leanspec/web');
  if (existsSync(webInParentNodeModules)) {
    return webInParentNodeModules;
  }
  
  // In development, might be in monorepo structure
  const webInMonorepo = resolve(currentDir, '../../web');
  if (existsSync(webInMonorepo)) {
    return webInMonorepo;
  }
  
  return null;
}

/**
 * Main entry point
 */
async function main() {
  const options = parseArgs(process.argv.slice(2));
  const cwd = process.cwd();

  // Determine specs directory
  let specsDir;
  if (options.specs) {
    specsDir = resolve(cwd, options.specs);
  } else {
    specsDir = detectSpecsDirectory(cwd);
  }

  // Verify specs directory exists
  if (!specsDir || !existsSync(specsDir)) {
    console.error(`✗ Specs directory not found${specsDir ? `: ${specsDir}` : ''}`);
    console.log('\nTried looking in:');
    console.log('  - ./specs');
    console.log('  - ./spec');
    console.log('  - ./docs/specs');
    console.log('  - ./docs/spec');
    console.log('  - ./.lean-spec/specs');
    console.log('  - leanspec.yaml config file');
    console.log('\nRun `npx lean-spec init` to initialize LeanSpec in this directory.');
    console.log('Or use --specs <dir> to specify a custom location.');
    process.exit(1);
  }

  // Validate port
  const port = validatePort(options.port);

  // Find web package
  const webDir = findWebPackage();
  if (!webDir) {
    console.error('✗ Could not find @leanspec/web package');
    console.log('\nMake sure @leanspec/web is installed:');
    console.log('  npm install @leanspec/web');
    process.exit(1);
  }

  console.log('→ Starting LeanSpec web UI...\n');
  console.log(`  Specs: ${specsDir}`);
  console.log(`  Port:  ${port}`);
  console.log(`  Open:  ${options.open ? 'yes' : 'no'}\n`);

  // Set environment variables for the web server
  const env = {
    ...process.env,
    SPECS_MODE: 'filesystem',
    SPECS_DIR: specsDir,
    PORT: port.toString(),
  };

  // Detect package manager in web directory
  const packageManager = existsSync(join(webDir, '../../pnpm-lock.yaml')) ? 'pnpm' :
                         existsSync(join(webDir, '../../yarn.lock')) ? 'yarn' : 'npm';

  // Start the Next.js dev server
  const child = spawn(packageManager, ['run', 'dev'], {
    cwd: webDir,
    stdio: 'inherit',
    env,
  });

  // Wait for server to be ready
  setTimeout(async () => {
    console.log(`\n✨ LeanSpec UI: http://localhost:${port}\n`);
    console.log('Press Ctrl+C to stop\n');

    if (options.open) {
      try {
        // Dynamic import of open package
        const openModule = await import('open');
        const open = openModule.default;
        await open(`http://localhost:${port}`);
      } catch (error) {
        // If open package not available, just show the URL
        console.log('⚠ Could not open browser automatically');
        console.log('Please visit the URL above manually\n');
      }
    }
  }, 5000);

  // Handle shutdown gracefully
  const sigintHandler = () => {
    child.kill('SIGTERM');
    console.log('\n✓ Web UI stopped');
    process.exit(0);
  };
  process.once('SIGINT', sigintHandler);

  // Handle child process exit
  child.on('exit', (code) => {
    if (code !== 0 && code !== null) {
      console.error(`\nProcess exited with code ${code}`);
      process.exit(code);
    }
  });

  child.on('error', (error) => {
    console.error(`\n✗ Error: ${error.message}`);
    process.exit(1);
  });
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: rust/leanspec-cli/src/commands/ui.rs
//! UI command implementation
//!
//! Start local web UI for spec management.

use colored::Colorize;
use std::error::Error;
use std::process::{Command, Stdio};
use std::path::Path;
use std::fs;

pub fn run(
    specs_dir: &str,
    port: &str,
    no_open: bool,
    multi_project: bool,
    dev: bool,
    dry_run: bool,
) -> Result<(), Box<dyn Error>> {
    // Validate port - parse to check if it's a valid number
    let _port_num: u16 = port.parse()
        .map_err(|_| format!("Invalid port number: {}", port))?;
    
    // Port validation happens during parse - u16 range is 0-65535
    
    let cwd = std::env::current_dir()?;
    let specs_path = cwd.join(specs_dir);
    
    // Verify specs directory exists (for single-project mode)
    if !multi_project && !specs_path.exists() {
        return Err(format!(
            "Specs directory not found: {}\n\nRun `lean-spec init` to initialize LeanSpec in this directory.",
            specs_path.display()
        ).into());
    }
    
    if multi_project {
        println!("{}", "→ Multi-project mode enabled".cyan());
    }
    
    // Check if we're in the LeanSpec monorepo for dev mode
    if dev {
        let ui_dir = cwd.join("packages/ui");
        let ui_package_json = ui_dir.join("package.json");
        
        if !ui_package_json.exists() {
            return Err("Development mode only works in the LeanSpec monorepo.\nRemove --dev flag to use production mode.".into());
        }
        
        // Check if it's the @leanspec/ui package
        let package_json_content = fs::read_to_string(&ui_package_json)?;
        if !package_json_content.contains("\"name\": \"@leanspec/ui\"") {
            return Err("Development mode only works in the LeanSpec monorepo.\nRemove --dev flag to use production mode.".into());
        }
        
        return run_dev_mode(&ui_dir, specs_dir, port, !no_open, multi_project, dry_run);
    }
    
    // Production mode: use published @leanspec/ui
    run_published_ui(&cwd, specs_dir, port, !no_open, multi_project, dry_run)
}

fn run_dev_mode(
    ui_dir: &Path,
    specs_dir: &str,
    port: &str,
    open_browser: bool,
    multi_project: bool,
    dry_run: bool,
) -> Result<(), Box<dyn Error>> {
    println!("{}\n", "→ Detected LeanSpec monorepo, using local ui package".dimmed());
    
    // Detect package manager
    let package_manager = detect_package_manager(ui_dir)?;
    
    let specs_mode = if multi_project { "multi-project" } else { "filesystem" };
    
    if dry_run {
        println!("{}", "Would run:".cyan());
        println!("  cd {}", ui_dir.display().to_string().dimmed());
        println!(
            "  SPECS_MODE={} SPECS_DIR={} PORT={} {} run dev",
            specs_mode, specs_dir, port, package_manager
        );
        if open_browser {
            println!("  open http://localhost:{}", port);
        }
        return Ok(());
    }
    
    println!("{}", "Starting web UI...".cyan());
    
    // Set environment variables
    let child = Command::new(&package_manager)
        .args(["run", "dev"])
        .current_dir(ui_dir)
        .env("SPECS_MODE", specs_mode)
        .env("SPECS_DIR", specs_dir)
        .env("PORT", port)
        .stdin(Stdio::inherit())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .spawn()?;
    
    println!();
    println!("{}", format!("✨ LeanSpec UI: http://localhost:{}", port).green());
    println!();
    println!("{}", "Press Ctrl+C to stop".dimmed());
    
    if open_browser {
        open_url(&format!("http://localhost:{}", port));
    }
    
    // Wait for the process
    let output = child.wait_with_output()?;
    
    if !output.status.success() {
        return Err("Web UI process exited with error".into());
    }
    
    Ok(())
}

fn run_published_ui(
    cwd: &Path,
    specs_dir: &str,
    port: &str,
    open_browser: bool,
    multi_project: bool,
    dry_run: bool,
) -> Result<(), Box<dyn Error>> {
    println!("{}\n", "→ Using published @leanspec/ui package".dimmed());
    
    // Detect package manager
    let package_manager = detect_package_manager(cwd)?;
    
    // Build command
    let (cmd, args) = build_ui_command(&package_manager, specs_dir, port, open_browser, multi_project);
    
    if dry_run {
        println!("{}", "Would run:".cyan());
        println!("  {} {}", cmd, args.join(" "));
        return Ok(());
    }
    
    let mut child = Command::new(&cmd)
        .args(&args)
        .current_dir(cwd)
        .stdin(Stdio::inherit())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .spawn()?;
    
    // Wait for the process
    let status = child.wait()?;
    
    if !status.success() {
        let code = status.code().unwrap_or(1);
        eprintln!();
        eprintln!("{}", format!("@leanspec/ui exited with code {}", code).red());
        eprintln!("{}", "Make sure npm can download @leanspec/ui.".dimmed());
        return Err("Web UI process exited with error".into());
    }
    
    Ok(())
}

fn build_ui_command(
    package_manager: &str,
    specs_dir: &str,
    port: &str,
    open_browser: bool,
    multi_project: bool,
) -> (String, Vec<String>) {
    let mut ui_args = vec!["@leanspec/ui".to_string()];
    
    if multi_project {
        ui_args.push("--multi-project".to_string());
    } else {
        ui_args.push("--specs".to_string());
        ui_args.push(specs_dir.to_string());
    }
    
    ui_args.push("--port".to_string());
    ui_args.push(port.to_string());
    
    if !open_browser {
        ui_args.push("--no-open".to_string());
    }
    
    match package_manager {
        "pnpm" => ("pnpm".to_string(), [vec!["dlx".to_string()], ui_args].concat()),
        "yarn" => ("yarn".to_string(), [vec!["dlx".to_string()], ui_args].concat()),
        _ => ("npx".to_string(), [vec!["--yes".to_string()], ui_args].concat()),
    }
}

fn detect_package_manager(dir: &Path) -> Result<String, Box<dyn Error>> {
    // Check for lockfiles
    if dir.join("pnpm-lock.yaml").exists() {
        return Ok("pnpm".to_string());
    }
    if dir.join("yarn.lock").exists() {
        return Ok("yarn".to_string());
    }
    if dir.join("package-lock.json").exists() {
        return Ok("npm".to_string());
    }
    
    // Check parent directories
    let mut current = dir.to_path_buf();
    while let Some(parent) = current.parent() {
        if parent.join("pnpm-lock.yaml").exists() {
            return Ok("pnpm".to_string());
        }
        if parent.join("yarn.lock").exists() {
            return Ok("yarn".to_string());
        }
        if parent.join("package-lock.json").exists() {
            return Ok("npm".to_string());
        }
        current = parent.to_path_buf();
    }
    
    // Default to npm
    Ok("npm".to_string())
}

fn open_url(url: &str) {
    #[cfg(target_os = "macos")]
    let _ = Command::new("open").arg(url).spawn();
    
    #[cfg(target_os = "linux")]
    let _ = Command::new("xdg-open").arg(url).spawn();
    
    #[cfg(target_os = "windows")]
    let _ = Command::new("cmd").args(["/C", "start", url]).spawn();
}
